### PR TITLE
Fix hardware assignment race by adding synchronization in ensureHardware

### PR DESCRIPTION
**Pull request by ptrivedi**
This change addresses a race condition in CAPT where the same hardware could get assigned to different nodes.

Reconcile process starts up as many worker threads as dictated by the
MaxConcurrentReconciles parameter of controller-runtime for each object/crd
that needs reconciliation. This calls for some kind of synchronization for
hardware assignment to machines. Hardware ownership is indicated by
adding/updating the ownerName/space labels on the hardware CRD. This warrants
that seeking for an available hardware and updating these labels must happen
atomically.

While this patch adds a lock around the ensureHardware function in the reconciler,
we may need further discussion on whether entire ensureDependencies would
warrant synchronization. There is the chance of another race where multiple
workflows could get created for the same hardware. This may not cause any
visible issues, except maybe cleanup/delete could leave stray workflows around.

## Why is this needed

Fixes: #

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

File: controllers/machine.go [hunk 1]
```diff
@@ -327,6 +331,12 @@ func (mrc *machineReconcileContext) ensureHardwareUserData(hardware *tinkv1.Hard
 }
 
 func (mrc *machineReconcileContext) ensureHardware() (*tinkv1.Hardware, error) {
+        // Make the process of picking up the hardware and updating it's ownership atomic.
+        // This avoids a race that could potentially cause picking up the same hardware 
+        // for different machines.
+        lock.Lock()
+        defer lock.Unlock()
```
Comment (invidian) [inline controllers/machine.go:338]: The indentation here is not aligned with rest of the function body.

File: controllers/machine.go [hunk 2]
```diff
@@ -327,6 +331,12 @@ func (mrc *machineReconcileContext) ensureHardwareUserData(hardware *tinkv1.Hard
 }
 
 func (mrc *machineReconcileContext) ensureHardware() (*tinkv1.Hardware, error) {
+        // Make the process of picking up the hardware and updating it's ownership atomic.
```
Comment (invidian) [inline controllers/machine.go:334]: Is the reconciler doing the leader election, so lock per instance is also ensured?

File: controllers/machine.go (hunk 1 above)
Comment (invidian) [inline controllers/machine.go:338]: Perhaps lock could be released right after `takeHardwareOwnership` to make time of the locking a bit shorter?

**Review by invidian — COMMENTED** [review]
The approach with lock seems reasonable to me, assuming that reconciler is already doing the leader election, so there is no 2 reconciler running simultaneously, which could make the lock weaker.

Perhaps `func machineReconciliationSelectsUniqueAndAvailablehardwareForEachMachine(t *testing.T) {` test should be hardened to reconcile many machines at the same time and then verify that each of them gets unique hardware assigned to ensure this lock works as expected and that the regression to this part of code is less likely in the future.

File: controllers/machine.go (hunk 2 above)
Comment (ptrivedi) [reply controllers/machine.go:334]: yes

File: controllers/machine.go (hunk 1 above)
Comment (ptrivedi) [reply controllers/machine.go:338]: Exactly what I had in mind and had tested with. I will have to try this again to recall exactly what the issue was, but the best way to fix this was to ensure atomicity of entire hardware update operation, including the userdata setting and patching via setStatus.

**Pull request closed without merging**